File: src/ClassFile/ClassFile.cpp
//
// Created by Danil on 27.08.2024.
//

#include "ClassFile/ClassFile.h"

ClassFile ClassFile::fromStream(std::istream &input) {
    ClassFile cf{};
    cf.magic = read_u4(input);
    cf.minor_version = read_u2(input);
    cf.major_version = read_u2(input);

    u2 constant_pool_count = read_u2(input);
    cf.constant_pool.reserve(constant_pool_count - 1); // !
    for (int i = 0; i < constant_pool_count - 1; ++i) // !!
        cf.constant_pool.push_back(cp_info::fromStream(input));

    cf.access_flags = read_u2(input);
    cf.this_class = read_u2(input);
    cf.super_class = read_u2(input);

    u2 interfaces_count = read_u2(input);
    cf.interfaces.reserve(interfaces_count);
    for (int i = 0; i < interfaces_count; ++i)
        cf.interfaces.push_back(read_u2(input));

    u2 fields_count = read_u2(input);
    cf.fields.reserve(fields_count);
    for (int i = 0; i < fields_count; ++i)
        cf.fields.push_back(field_info::fromStream(input));

    u2 methods_count = read_u2(input);
    cf.methods.reserve(methods_count);
    for (int i = 0; i < methods_count; ++i)
        cf.methods.push_back(method_info::fromStream(input));

    u2 attributes_count = read_u2(input);
    cf.attributes.reserve(attributes_count);
    for (int i = 0; i < attributes_count; ++i)
        cf.attributes.push_back(attribute_info::fromStream(input));

    return cf;
}

void ClassFile::saveToStream(std::ostream &output) const {
    write(output, magic);
    write(output, minor_version);
    write(output, major_version);

    write(output, (u2) (constant_pool.size() + 1));
    for (const auto &info: constant_pool)
        info.saveToStream(output);

    write(output, this->access_flags);
    write(output, this->this_class);
    write(output, this->super_class);

    write(output, (u2) this->interfaces.size());
    for (const auto &interface: this->interfaces)
        write(output, interface);

    write(output, (u2) this->fields.size());
    for (const auto &field: this->fields)
        field.saveToStream(output);

    write(output, (u2) this->methods.size());
    for (const auto &method: this->methods)
        method.saveToStream(output);

    write(output, (u2) this->attributes.size());
    for (const auto &attribute: this->attributes)
        attribute.saveToStream(output);
}

File: src/class_file/class_file.rs
use std::io::{Read, Write};
use byteorder::{ReadBytesExt, WriteBytesExt};
use crate::class_file::attribute_info::AttributeInfo;
use crate::class_file::constant_pool::CpInfo;
use crate::class_file::class_file_error::ClassFileError;
use crate::class_file::field_info::FieldInfo;
use crate::class_file::method_info::MethodInfo;

#[derive(Debug, PartialEq)]
pub struct ClassFile {
    magic: u32,
    minor_version: u16,
    major_version: u16,
    constant_pool: Vec<CpInfo>,
    access_flags: u16,
    this_class: u16,
    super_class: u16,

    interfaces: Vec<u16>,
    fields: Vec<FieldInfo>,
    methods: Vec<MethodInfo>,
    attributes: Vec<AttributeInfo>,
}

impl ClassFile {
    pub fn read<R: Read>(reader: &mut R) -> Result<ClassFile, ClassFileError> {
        let magic = reader.read_u32::<byteorder::BigEndian>()?;
        let minor_version = reader.read_u16::<byteorder::BigEndian>()?;
        let major_version = reader.read_u16::<byteorder::BigEndian>()?;
        let constant_pool_count = reader.read_u16::<byteorder::BigEndian>()?;
        let constant_pool = std::iter::repeat_with(|| CpInfo::read(reader))
            .take((constant_pool_count - 1) as usize)
            .collect::<Result<Vec<CpInfo>, ClassFileError>>()?;
        let access_flags = reader.read_u16::<byteorder::BigEndian>()?;
        let this_class = reader.read_u16::<byteorder::BigEndian>()?;
        let super_class = reader.read_u16::<byteorder::BigEndian>()?;
        let interfaces_count = reader.read_u16::<byteorder::BigEndian>()?;
        let interfaces = std::iter::repeat_with(|| reader.read_u16::<byteorder::BigEndian>())
            .take(interfaces_count as usize)
            .collect::<Result<Vec<u16>, std::io::Error>>()?;
        let fields_count = reader.read_u16::<byteorder::BigEndian>()?;
        let fields = std::iter::repeat_with(|| FieldInfo::read(reader))
            .take(fields_count as usize)
            .collect::<Result<Vec<FieldInfo>, ClassFileError>>()?;

        let methods_count = reader.read_u16::<byteorder::BigEndian>()?;
        let methods = std::iter::repeat_with(|| MethodInfo::read(reader))
            .take(methods_count as usize)
            .collect::<Result<Vec<MethodInfo>, ClassFileError>>()?;

        let attributes_count = reader.read_u16::<byteorder::BigEndian>()?;
        let attributes = std::iter::repeat_with(|| AttributeInfo::read(reader))
            .take(attributes_count as usize)
            .collect::<Result<Vec<AttributeInfo>, ClassFileError>>()?;

        Ok(ClassFile {
            magic,
            minor_version,
            major_version,
            constant_pool,
            access_flags,
            this_class,
            super_class,
            interfaces,
            fields,
            methods,
            attributes,
        })
    }

    pub fn write<W: Write>(&self, writer: &mut W) -> Result<(), std::io::Error> {
        writer.write_u32::<byteorder::BigEndian>(self.magic)?;
        writer.write_u16::<byteorder::BigEndian>(self.minor_version)?;
        writer.write_u16::<byteorder::BigEndian>(self.major_version)?;
        writer.write_u16::<byteorder::BigEndian>((self.constant_pool.len() + 1) as u16)?;
        for constant_pool in &self.constant_pool {
            constant_pool.write(writer)?;
        }
        writer.write_u16::<byteorder::BigEndian>(self.access_flags)?;
        writer.write_u16::<byteorder::BigEndian>(self.this_class)?;
        writer.write_u16::<byteorder::BigEndian>(self.super_class)?;
        writer.write_u16::<byteorder::BigEndian>(self.interfaces.len() as u16)?;
        for interface in &self.interfaces {
            writer.write_u16::<byteorder::BigEndian>(*interface)?;
        }
        writer.write_u16::<byteorder::BigEndian>(self.fields.len() as u16)?;
        for field in &self.fields {
            field.write(writer)?;
        }
        writer.write_u16::<byteorder::BigEndian>(self.methods.len() as u16)?;
        for method in &self.methods {
            method.write(writer)?;
        }
        writer.write_u16::<byteorder::BigEndian>(self.attributes.len() as u16)?;
        for attribute in &self.attributes {
            attribute.write(writer)?;
        }
        Ok(())
    }
}
